### PR TITLE
[RFC] Discrete distributions

### PR DESCRIPTION
**Pull request by regislebrun**
This is an almost finished work on discrete distributions. It includes:
+ more accurate and efficient algorithms for Binomial and Poisson PDF evaluation (Loader's algorithm)
+ improved Multinomial distribution
+ univariate and multivariate Hypergeometric distributions

Thanks for the feedback.

### CODE REVIEW

File: lib/src/Uncertainty/Distribution/Multinomial.cxx [hunk 1]
```diff
@@ -136,49 +136,86 @@ Point Multinomial::getRealization() const
...
+  }
   for (UnsignedInteger i = 0; i < dimension; ++i)
   {
-    const Scalar k = point[i];
-    // For p_[i] > 0, it is possible to obtain any value of k between 0 and n
-    if (p_[i] > 0.0) logPDF += k * std::log(p_[i]) - lgamma(k + 1.0);
-    // Else only k == 0 is allowed, with a zero contribution to the log PDF
-    else if (k > 0) return 0.0;
+    // Here we know that round(point[i]) >= 0
+    const UnsignedInteger k = static_cast<UnsignedInteger>(round(point[i]));
```
Comment (jschueller) [inline lib/src/Uncertainty/Distribution/Multinomial.cxx]: k => warning: declaration shadows a local variable [-Wshadow]

File: lib/src/Uncertainty/Distribution/Multinomial.cxx [hunk 2]
```diff
@@ -310,6 +364,88 @@ Scalar Multinomial::computeCDF(const Point & point) const
...
+  }
+  const Point lower(interval.getLowerBound());
+  const Point upper(interval.getUpperBound());
+  Indices a(dimension_);
+  Indices b(dimension_);
+  UnsignedInteger sigmaA = 0;
+  UnsignedInteger sigmaB = 0;
+  for (UnsignedInteger i = 0; i < dimension_; ++i)
+    {
+      a[i] = std::max(0UL, static_cast<UnsignedInteger>(std::ceil(lower[i])));
```
Comment (jschueller) [inline lib/src/Uncertainty/Distribution/Multinomial.cxx]: max(0, uint x) will always return x

File: lib/src/Uncertainty/Distribution/Multinomial.cxx [hunk 3]
```diff
@@ -310,6 +364,88 @@ Scalar Multinomial::computeCDF(const Point & point) const
...
+	break;
+      const Complex zeta(r * std::exp(Complex(0.0, k * M_PI / nA)));
+      delta = sign2 * (computeGlobalPhi(zeta, a, b) * std::exp(Complex(0.0, k * sigmaA * M_PI / nA))).real();
+      value += delta;
+      sign2 = -sign2;
+    }
+  // Check if we have to take the last term into account
+  if (std::abs(delta) > SpecFunc::Precision * std::abs(value))
+    {
+      const Scalar delta = coefNorm * computeGlobalPhi(-r, a, b).real();
```
Comment (jschueller) [inline lib/src/Uncertainty/Distribution/Multinomial.cxx]: delta => declaration shadows a local variable [-Wshadow]

File: lib/src/Uncertainty/Distribution/Multinomial.cxx (hunk 1 above)
Comment (sofianehaddad) [reply lib/src/Uncertainty/Distribution/Multinomial.cxx]: indeed k declared first as Scalar then UI...

File: lib/src/Uncertainty/Distribution/Hypergeometric.cxx [hunk 4]
```diff
@@ -0,0 +1,443 @@
...
+Hypergeometric::Hypergeometric()
+  : DiscreteDistribution()
+  , n_(0)
+  , k_(0)
+  , m_(0)
+{
+  setName( "Hypergeometric" );
+  // We set the dimension of the Hypergeometric distribution
+  setDimension( 1 );
+  computeProbabilities();
```
Comment (jschueller) [inline lib/src/Uncertainty/Distribution/Hypergeometric.cxx]: The default parametrization n=0 k=0 m=0 doesn't yield a valid distribution and crashes in computeProbabilities at line 355 : `probabilities_[kMode] = p;`.

File: lib/src/Uncertainty/Distribution/MultivariateHypergeometric.cxx [hunk 5]
```diff
@@ -0,0 +1,836 @@
...
+
+typedef Collection<UnsignedInteger>     UnsignedIntegerCollection;
+
+CLASSNAMEINIT(Multinomial)
+
+static const Factory<Multinomial> Factory_Multinomial;
+
+/* Default constructor */
+Multinomial::Multinomial()
+  : DiscreteDistribution()
```
Comment (jschueller) [inline lib/src/Uncertainty/Distribution/MultivariateHypergeometric.cxx]: This is not ready, leave it for the next release.

File: lib/src/Base/Func/SpecFunc/SpecFunc.cxx [hunk 6]
```diff
@@ -745,6 +769,36 @@ Scalar SpecFunc::TriGamma(const Scalar x)
...
+  static Scalar S2 = 7.93650793650793651e-04;
+  static Scalar S3 = 5.95238095238095238e-04;
+  static Scalar S4 = 8.41750841750841751e-04;
+  const Scalar nn = (1.0 * n) * n;
+  if (n > 2559) return (S0 - S1 / nn) / n;
+  if (n > 82)  return (S0 - (S1 - S2 / nn) / nn) / n;
+  if (n > 50)  return (S0 - (S1 - (S2 - S3 / nn) / nn) / nn) / n;
+  return (S0 - (S1 - (S2 - (S3 - S4 / nn) / nn) / nn) / nn) / n;
+}
+
```
Comment (jschueller) [inline lib/src/Base/Func/SpecFunc/SpecFunc.cxx]: for reference, where did you get this ?

I tried to look elsewhere for comparison, but I did not find the same implementation:
https://kaskr.github.io/adcomp/stirlerr_8cpp_source.html

File: lib/src/Base/Func/SpecFunc/openturns/SpecFunc.hxx [hunk 7]
```diff
@@ -217,6 +220,8 @@ public:
   static Scalar DiGammaInv(const Scalar a);
   // Trigamma function: TriGamma(x) = ((d^2gamma/dx^2) / gamma)(x)
   static Scalar TriGamma(const Scalar x);
+  // Stirling error: stirlerr(n) = log(n!) - log( sqrt(2*pi*n)*(n/e)^n )
+  static Scalar stirlerr(const UnsignedInteger n);
```
Comment (jschueller) [inline lib/src/Base/Func/SpecFunc/openturns/SpecFunc.hxx]: static, so an uppercase is needed: Stirlerr

File: lib/src/Base/Type/openturns/Point.hxx [hunk 8]
```diff
@@ -55,6 +55,9 @@ public:
   /** Constructor from a collection */
   Point(const Collection<Scalar> & coll);
 
+  /** Constructor from a collection */
+  explicit Point(const Collection<UnsignedInteger> & coll);
```
Comment (jschueller) [inline lib/src/Base/Type/openturns/Point.hxx]: I'm not sure about this one where do you need it ?

Comment (sofianehaddad) [reply lib/src/Base/Type/openturns/Point.hxx]: In https://github.com/openturns/openturns/pull/1088/commits/3c1bafc2817ac1154019864e0339ed2dd833c0f2

Comment (sofianehaddad) [reply lib/src/Base/Type/openturns/Point.hxx]: And a I agree I think it should be dropped

File: lib/src/Base/Func/SpecFunc/SpecFunc.cxx [hunk 9]
```diff
@@ -526,6 +526,30 @@ Scalar SpecFunc::FaddeevaIm(const Scalar x)
   return Faddeeva::w_im(x);
 }
 
+// Factorial function
+Scalar SpecFunc::LogFactorial(const UnsignedInteger n)
```
Comment (jschueller) [inline lib/src/Base/Func/SpecFunc/SpecFunc.cxx]: The new functions must have an entry in user_manual/functions.rst

File: lib/src/Uncertainty/Distribution/Multinomial.cxx [hunk 10]
```diff
@@ -233,38 +294,31 @@ Scalar Multinomial::computeCDF(const Point & point) const
...
+  const Indices kPoint(point.begin(), point.end());
   // Early exit for 1D case
   if (dimension == 1)
   {
-    const Scalar k = point[0];
-    if (k < -supportEpsilon_) return 0.0;
-    if (k > n_ + supportEpsilon_) return 1.0;
-    return DistFunc::pBeta(n_ - floor(k), floor(k) + 1, 1.0 - p_[0]);
+    if (point[0] < -supportEpsilon_) return 0.0;
+    if (point[1] > n_ + supportEpsilon_) return 1.0;
```
Comment (jschueller) [inline lib/src/Uncertainty/Distribution/Multinomial.cxx]: I've got a crash here, it should be point[0]

File: python/src/Hypergeometric_doc.i.in [hunk 11]
```diff
@@ -0,0 +1,105 @@
...
+-------
+m : int, :math:`0\leq m\leq n`
+    The sample size."
+
+// ---------------------------------------------------------------------
+
+%feature("docstring") OT::Hypergeometric::getN
+"Accessor to the poppulation size.
+
+Parameters
```
Comment (jschueller) [inline python/src/Hypergeometric_doc.i.in]: Returns

File: python/src/Hypergeometric_doc.i.in [hunk 12]
```diff
@@ -0,0 +1,105 @@
...
+
+Returns
+-------
+m : int, :math:`0\leq m\leq n`
+    The sample size."
+
+// ---------------------------------------------------------------------
+
+%feature("docstring") OT::Hypergeometric::getN
+"Accessor to the poppulation size.
```
Comment (jschueller) [inline python/src/Hypergeometric_doc.i.in]: population

File: python/doc/bibliography.rst [hunk 13]
```diff
@@ -43,6 +43,13 @@ Bibliography
 .. [devroye1986] Devroye L, *Non-Uniform RandomVariate Generation*,
     Springer-Verlag, New York, 1986
     `pdf <http://doc.openturns.org/papers/devroye1986.pdf>`__
+.. [devroye1986b] Devroye L, *Non-Uniform RandomVariate Generation - Errata*,
```
Comment (jschueller) [inline python/doc/bibliography.rst]: this one is not used, so we'll get an error:
`bibliography.rst:46:Citation [devroye1986b] is not referenced.`

File: python/doc/bibliography.rst [hunk 14]
```diff
@@ -54,9 +61,10 @@ Bibliography
 .. [fang2006] K-T. Fang, R. Li, and A. Sudjianto. *Design and modeling for
     computer experiments.* Chapman & Hall CRC, 2006.
 .. [gamboa2013] Gamboa, F., Janon, A., Klein, T. & Lagnoux, A. *Sensitivity
-    analysis for multidimensional and functional outputs.* 2013
+    analysis for multidimensional and functional outputs.* 2013.
     `pdf <http://doc.openturns.org/papers/gamboa2013.pdf>`__
-.. [hormann1992] Hormann W., *The generation of Binomial Random Variates*
```
Comment (jschueller) [inline python/doc/bibliography.rst]: theory/numerical_methods/distribution_realization.rst still references hormann1992

File: python/src/Hypergeometric_doc.i.in [hunk 15]
```diff
@@ -0,0 +1,105 @@
...
+    Size of the sample drawn in the population.
+
+Notes
+-----
+Its probability density function is defined as:
+
+.. math::
+
+    \Prob{X = i} = \dfrac{\binom{k, i}\binom{n-m, m-i}}{\binom{n, m}}
+                   \quad \forall i \in \{\max\{0,k+m-n\}, \ldots, \min\{k,m\}\}
```
Comment (jschueller) [inline python/src/Hypergeometric_doc.i.in]: seems this equation yields an issue:
```
! Argument of \@genfrac has an extra }.
<inserted text> 
                \par 
l.17 ...,k+m-n\}, \ldots, \min\{k,m\}\}\end{split}
                                                  
Runaway argument?
! Paragraph ended before \@genfrac was complete.
```

File: python/src/DistFunc_doc.i.in [hunk 16]
```diff
@@ -210,3 +294,85 @@ Examples
...
+
+Returns
+-------
+realizations : int or :class:`~openturns.Indices`
+    The realizations of the discrete disctribution.
+
+Notes
+-----
+For the small values of :math:`\lambda`, we use the method of inversion by
+sequential search described in [devroy1986]_ and with the important errata in
```
Comment (jschueller) [inline python/src/DistFunc_doc.i.in]: devroye1986

File: python/src/DistFunc_doc.i.in [hunk 17]
```diff
@@ -210,3 +294,85 @@ Examples
...
+[devroy1986b]_. For the large values of :math:`\lambda`, we use the ratio of
```
Comment (jschueller) [inline python/src/DistFunc_doc.i.in]: devroye1986b

File: lib/src/Uncertainty/Distribution/Multinomial.cxx (hunk 1 above)
Comment (regislebrun) [reply lib/src/Uncertainty/Distribution/Multinomial.cxx]: k is declared as a const UnsignedInteger in a limited scope (line 182) then as an UnsignedInteger (line 194) then as a const UnsignedInteger in a limited scope (line 205), where the shadowing occurs. It was a Scalar in the previous version of the code.
I fixed it by declaring k as an UnsignedInteger once and for all before the first loop.

File: lib/src/Uncertainty/Distribution/Multinomial.cxx (hunk 2 above)
Comment (regislebrun) [reply lib/src/Uncertainty/Distribution/Multinomial.cxx]: Yep. Cast and max should be reversed. Done.

File: python/src/DistFunc_doc.i.in (hunk 16 above)
Comment (regislebrun) [reply python/src/DistFunc_doc.i.in]: fixed

File: python/src/DistFunc_doc.i.in (hunk 17 above)
Comment (regislebrun) [reply python/src/DistFunc_doc.i.in]: Fixed. Thanks, this error drove me crazy!

File: python/doc/bibliography.rst (hunk 14 above)
Comment (regislebrun) [reply python/doc/bibliography.rst]: Fixed

File: python/src/Hypergeometric_doc.i.in (hunk 15 above)
Comment (regislebrun) [reply python/src/Hypergeometric_doc.i.in]: Strange. Will see for an alternative.

File: python/src/Hypergeometric_doc.i.in (hunk 12 above)
Comment (regislebrun) [reply python/src/Hypergeometric_doc.i.in]: Done

File: python/src/Hypergeometric_doc.i.in (hunk 11 above)
Comment (regislebrun) [reply python/src/Hypergeometric_doc.i.in]: Done

File: lib/src/Base/Func/SpecFunc/SpecFunc.cxx (hunk 6 above)
Comment (regislebrun) [reply lib/src/Base/Func/SpecFunc/SpecFunc.cxx]: The formula given in Loader's paper does not provide full double precision. I had to increase the threshold a little bit to achieve the needed precision and to increase the lookup table accordingly.

File: lib/src/Base/Type/openturns/Point.hxx (hunk 8 above)
Comment (regislebrun) [reply lib/src/Base/Type/openturns/Point.hxx]: Ok

**Comment by jschueller** [comment]
- there is still an issue with the devroye citation not being used
- pyinstallcheck_DistFunc_binomial fails for the last digits under mingw

**Comment by jschueller** [comment]
```
--- a/python/doc/user_manual/probabilistic_modelling.rst
+++ b/python/doc/user_manual/probabilistic_modelling.rst
@@ -259,9 +259,15 @@ Low-level distribution functions
     :toctree: _generated/
     :template: function.rst_t
 
+    DistFunc_dBinomial
+    DistFunc_dPoisson
+    DistFunc_logpBinomial
+    DistFunc_logpPoisson
     DistFunc_pNormal
     DistFunc_qNormal
+    DistFunc_rBinomial
     DistFunc_rNormal
+    DistFunc_rPoisson
     DistFunc_kFactor
     DistFunc_kFactorPooled
```

**Comment by jschueller** [comment]
Can you make sure to update it for tomorrow ?

**Comment by regislebrun** [comment]
@jschueller It looks like continuous integration is broken. Grrr.

**Comment by jschueller** [comment]
nah, it should be fixed now
.
I see the doc passes now, but there still some doctest issue, and a vs2010 build error

File: python/src/DistFunc_doc.i.in [hunk 18]
```diff
@@ -26,6 +112,131 @@ Examples
...
+
+Notes
+-----
+This method is based on the alias method.
+
+Examples
+--------
+>>> import openturns as ot
+>>> ot.RandomGenerator.SetSeed(0)
+>>> r = ot.DistFunc.rHypergeometric(10, 4, 7, 0.3)
```
Comment (jschueller) [inline python/src/DistFunc_doc.i.in]: this will still fail

**Comment by jschueller** [comment]
there's one last failing test on mingw:
```
-dBinomial( 5 , 0.9 ,  4 )=0.328
+dBinomial( 5 , 0.9 ,  4 )=0.3281
```

File: python/src/DistFunc_doc.i.in [hunk 19]
```diff
@@ -26,6 +112,131 @@ Examples
...
+    Flag to tell if it is the CDF or its complement which is evaluated
+
+Returns
+-------
+p : float
+    The probability to get at most :math:`x` candidates in a draw.
+
+Notes
+-----
+This method is based on a summation of the probability function toward the upper bound or the lower bound of the range depending on the position of :math:`x` wrt the mode :math:`\left\lfoor\dfrac{(k+1)(m+1)}{n+2}\right\rfloor` of the distribution, then take the complement if needed.
```
Comment (jschueller) [inline python/src/DistFunc_doc.i.in]: \lfoor -> \floor ?

Comment (regislebrun) [reply python/src/DistFunc_doc.i.in]: ok this time it should be good. Fixed the test, the LaTeX operator and the length of the lines in DistFunc_doc.i.in!

**Comment by jschueller** [comment]
unfortunately there are still some numeric issues with the binomial test

**Comment by regislebrun** [comment]
At this point it drives me crazy. Our testing procedures are pure crap: I am obliged to reduce the number of digits to 2 in order to pass the tests on a platform I cannot check locally, for a method supposed to provide almost full double precision. When I split the lines in the documentation it generates errors I cannot understand, and each correction/compilation/check takes more than one hour with this f*cking damn slow sphinx!

**Comment by jschueller** [comment]
If it helps, you can speed up sphinx with `cmake -DSPHINX_FLAGS="-j8"`

**Comment by jschueller** [comment]
I believe cppcheck_DistFunc_binomial is the last one to fix... one last time ?

**Comment by jschueller** [comment]
cppcheck_DistFunc_binomial still fails on mingw:
```
-dBinomial(5, 0.9, 4)=0.328
+dBinomial(5, 0.9, 4)=0.3281
```

**Comment by regislebrun** [comment]
Either this PR is merge as is or it is postponed to the next release. Tired to run after f*cking precision issues.

**Comment by jschueller** [comment]
ok, I'll finish it

**Pull request closed without merging**